### PR TITLE
update: Fire a pixel on install from ATB.

### PR DESCRIPTION
**Pull request by mrshu**
* Fire a pixel on install of the extension from the homepage.

Signed-off-by: mr.Shu <mr@shu.io>

### CODE REVIEW

File: lib/ddg-oninstall.js [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+var pageMod = require("sdk/page-mod");
+var pagemod;
+
+exports.install = function (is_install) {
+  pagemod = pageMod.PageMod({
+      include: /^https?:\/\/duckduckgo\.com\/?$/,
+      contentScriptFile: './js/duckduckgo.js',
+      attachTo: ["existing", "top"],
+      onAttach: function(worker) {
```
Comment (mrshu) [inline lib/ddg-oninstall.js]: @bsstoner This makes sure that the `js/duckduckgo.js` file is injected even if the tab already exists (look for "existing" on https://developer.mozilla.org/en-US/Add-ons/SDK/High-Level_APIs/page-mod)

File: lib/main.js [hunk 2]
```diff
@@ -121,12 +122,14 @@ exports.main = function(options, callbacks) {
         ss.storage.meanings = true;
     }
 
+    var is_install = false;
```
Comment (mrshu) [inline lib/main.js]: @bsstoner In order to test this out I suggest you change the value of this variable to `true` and open `https://duckduckgo.com` in the "test session". A pixel request should then be visible. I understand that this is not the same as the intended usecase but it is the closest we can get to, given the constrains.

File: lib/ddg-oninstall.js [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+ */
+
+"use strict";
+
+var pageMod = require("sdk/page-mod");
+var pagemod;
+
+exports.install = function (is_install) {
+  pagemod = pageMod.PageMod({
+      include: /^https?:\/\/duckduckgo\.com\/?$/,
```
Comment (bsstoner) [inline lib/ddg-oninstall.js]: I think we will want this to run on both the SERP and the homepage

File: lib/main.js [hunk 4]
```diff
@@ -205,6 +208,7 @@ exports.main = function(options, callbacks) {
       }
     }
     noatb.install();
+    oninstall.install(is_install);
```
Comment (bsstoner) [inline lib/main.js]: Will this be called every time the extension loads? Is there any benefit to only calling it if `is_install` is true?

File: lib/main.js (hunk 2 above)
Comment (bsstoner) [inline lib/main.js]: 👍 thanks, I can see it working doing that, makes sense

File: lib/ddg-oninstall.js (hunk 3 above)
Comment (mrshu) [inline lib/ddg-oninstall.js]: @bsstoner I see, I missed that the extension would be installable from the SERP too. Will fix.

File: lib/main.js (hunk 4 above)
Comment (mrshu) [inline lib/main.js]: @bsstoner There is the obvious benefit that the extension will not inject the script unnecessarily. I am not sure why I did not come up with this myself. Will change this too.

**Comment by mrshu** [comment]
@bsstoner thanks for your review, this is ready for another pass.

**Comment by bsstoner** [comment]
One last question, if they install this from the Mozilla store, the first time they navigate to duckduckgo.com will it send the pixel?

**Comment by mrshu** [comment]
@bsstoner not necessarily but it might happen sometimes.

As it stands currently, if you install this from the Mozilla store, the first time you navigate to duckduckgo.com *in the same session* (that is, until you close the browser window) a pixel will be sent.

I believe this is a bit different from the desired functionality -- that is to only send the pixel if there exist a tab with duckduckgo.com open in it. I believe we can achieve this by only letting the pixel to be sent in the first few (say 3) seconds after calling `install()`. If duckduckgo.com is already opened, it gives the extension plenty of time to inject the script and send the pixel. If not, it seems unlikely a user would open duckduckgo.com that quickly.

I understand that this still does not cover the case when the user installs the extension from the Mozilla store while having an unrelated duckduckgo.com tab opened. Maybe the `./js/duckduckgo.js` script could check whether the user clicked on the ATB modal or something like that? I am not sure how doable would that be, given all the restrictions.

What do you think? If you like the second option, I'll gladly go ahead an investigate some available options.

**Comment by bsstoner** [comment]
Thanks, that makes sense. The time restriction seems like it would be a good idea if you think it's doable without a bunch of extra code or overhead? I'm less concerned about the pixel and more thinking about if I add the success screen with animations and stuff it would be good to take as many precautions as possible to ensure that doesn't get shown to people unexpectedly.

**Comment by mrshu** [comment]
@bsstoner Cool, I've put together a timeout as described above. Moreover, this will happen on just for one tab open with duckduckgo.com in it. I am not sure if this is desirable now so please do let me know -- it's very easy to change.

**Comment by bsstoner** [comment]
@mrshu The code looks good. I don't totally follow what you mean by "this will happen on just for one tab open with duckduckgo.com in it"?

**Comment by mrshu** [comment]
@bsstoner Thanks to [this line](https://github.com/duckduckgo/firefox-zeroclickinfo/pull/48/files#diff-bfab4b127a48c0f4f2800018d18a07acR39) the pixel will not be sent more than once.

Does that help?

**Comment by bsstoner** [comment]
👍

**Pull request merged**